### ISSUE TITLE
All the JWTs

### ISSUE THREAD

**Issue by blainsmith** [oddnetworks/oddworks#214]
Wanted to reign in some confusion around the JWTs we have floating around Oddworks and hope to shed light how how we can use the claims correctly.

### Definitions
`iss` - Service that CREATED the token.
`sub` - Service REQUESTING access, can be the same as `iss`.
`aud` - List of services that `sub` is allowed to access with this token.

Unfortunately, these 3 claims do not provide ACL control to the services listed in `aud` like GET, POST, or any other arbitrary action.

### Oddworks Use Cases

- Developers need read-only tokens to ship in their channels to `GET /config`
- Developers need read-only viewer-based tokens to identify subsequent requests for analytics/auth
- Other Odd services need read and write access

If I am missing any then speak up.

#### Standardized `iss`

We can use the same value for all tokens indicating that the token was generated from within the Oddworks ecosystem.

```json
{
  "iss": "urn:oddworks"
}
```

#### `sub` Based on Service

The `sub` should indicate WHO/WHAT is requesting access. Here we are specifying a `device` is requesting access, but `"sub": "api.oddworks.io"` is also valid when that service wants access.

```json
{
  "iss": "urn:oddworks",
  "sub": "device"
}
```

#### `aud` Based on Service

Now we need to use `aud` to specify what services the `sub` is allowed access to with this particular token.

```json
{
  "iss": "urn:oddworks",
  "sub": "device",
  "aud": ["content.oddworks.io", "analytics.oddworks.io"]
}
```

Now the issue we haven't addressed is the actions a `sub` can do to the `aud`. Unfortunately, there is no native JWT claim that provides a spec for this. AWS has an interesting solution with their **policies**, but that is definitely overkill for now, but they are a full blown platform so it makes sense.

Since we now limit based on HTTP actions we can extent the `aud` to include those actions within the JWT for a simplistic approach. Using `*` gives you all actions. Not specifying only grants you reads like `GET`, `HEAD`, `OPTIONS`. 

```json
{
  "iss": "urn:oddworks",
  "sub": "device",
  "aud": ["content.oddworks.io::READ", "analytics.oddworks.io::POST"]
}
```

- `READ` includes `GET`, `HEAD`, `OPTIONS`
- `WRITE` includes `POST`, `PATCH`, `DELETE`
- Specify individual actions for custom access. (ex. `"content.oddworks.io::PATCH,DELETE`)

Not sure of the effect on Oddworks as it currently stands, but since ZERO apps have shipped now would be the time to make all the people happy and solidify a standard format based on the JWT spec and default claims for simplicity.

**Comment by pmcconnell** [comment]
This is a big help.

We need to document the payload requirements, if any, for each service. 

For example I know events needs channel, platform, viewer and aud where aud must contain one of 'odd-events-service' or 'platform' 

Obviously this can be massaged to meet this proposal but requirements should be noted.

**Comment by blainsmith** [comment]
Yes. Each service would need to have itself documented that says what **additional** payload keys are required for business logic once the JWT is validated. `odd-events-service` will allow the following JWT through the gates.

```json
{
  "iss": "urn:oddworks",
  "sub": "device",
  "aud": ["content.oddworks.io::READ", "analytics.oddworks.io::POST"]
}
```

However it will respond error the request since it doesn't contain the additional payload keys it needs. A fully valid JWT would be if the `odd-api-service` wants to post data for some reason.

```json
{
  "iss": "urn:oddworks",
  "sub": "api.oddworks.io",
  "aud": ["analytics.oddworks.io::POST"],
  "channel": "nasa",
  "platform": "roku",
  "viewer": "user12345"
}
```

**Comment by aljachimiak** [comment]
iss - Service that CREATED the token. **`iss`uer**
sub - Service REQUESTING access, can be the same as iss. **`sub`ject**
aud - List of services that sub is allowed to access with this token. **`aud`ience**

**Comment by aljachimiak** [comment]
What is `urn`?

**Comment by blainsmith** [comment]
`urn` stands for Universal Resource Name. I think @kixxauth picked that up from AWS.

**Comment by blainsmith** [comment]
I had some time to think about this and I think we should not place the HTTP actions within the tokens. In the event we want to grant more privileges to tokens we'd need to issue new ones which is bad when apps are shipped into the stores.

Instead we should let the implementing server decide what actions a token can perform. So now we are back to.

```json
// Standard "device" or "platform" or whatever name token for devs
{
  "iss": "urn:oddworks",
  "sub": "device",
  "aud": ["content.oddworks.io", "analytics.oddworks.io"],
  "channel": "nasa",
  "platform": "roku"
}

// Standard "viewer" token for devs
{
  "iss": "urn:oddworks",
  "sub": "device",
  "aud": ["content.oddworks.io", "analytics.oddworks.io"],
  "channel": "nasa",
  "platform": "roku",
  "viewer": "user12345"
}

// Our current "admin" token so api. can do anything it wants to content.
{
  "iss": "urn:oddworks",
  "sub": "api.oddworks.io",
  "aud": ["content.oddworks.io"]
}
```

**Comment by kixxauth** [comment]
Couple notes on JWT authorization:

We do have authorization white listing per HTTP method and endpoint in our request-authorization middleware: https://github.com/oddnetworks/oddworks/blob/master/lib/middleware/request-authorize.js

It currently makes no determination about what the "aud" (audience) means other than the fact that it allows the system to configure a whitelist for it.

**!HOWEVER** We have been using the strings "admin" to indicate an audience that has global write access, and "platform" to indicate an audience that does not have write access. This logic is reflected in a lot of the testing as well as the few implementations of Oddworks we have out there in production. So, we need to keep that in mind if we start playing with that.
